### DISCUSSION TITLE
ASP.NET MVC and YARP

### DISCUSSION THREAD

**Discussion by colin-sim** [dotnet/yarp#1464] [Q&A] [answered]
Is YARP suited for the following use case please?

We want to re-write a legacy web API in .NET 6 without having to do it in one big bang. We would like to start with a proxy application that will:
1. forward the request to the legacy application.
2. in a piecemeal way re-write the APIs in the proxy application (i.e. as ASP.NET controllers). The results from both will be compared and monitored over a period of time to ensure there are no differences.
3. replace legacy API with new one once we have confidence there is no change in behaviour.

I would assume this is a fairly common scenario. If YARP is a good framework for this use case are there any code samples or guidance on how this may be achieved please?

I have experimented by adding in a ASP.NET middleware that forwards the request to the legacy application using `IHttpForwarder`, copy the response then forward the request through the middleware pipeline; however, I've found that this approach:
1. isn't very efficient when requests/responses are large as they are cached in memory.
2. the ASP.NET pipeline does some *odd* things like sending the response back before I get a chance to compare the middleware's resonse.

**Comment by Tratcher** [comment, marked as answer]
The gradual migration is well suited for YARP, but the part about sending requests to both and comparing them is a bit more challenging. See https://github.com/microsoft/reverse-proxy/issues/105.

**Reply by pstatho** [reply]
My intention is to use YARP specifically for this and I'm pretty excited about migrating to .net6 in a piecemeal fashion. 

However I'm having a few issues with the initial setup. At the beginning only a small number of pages will be in .net6 and the vat majority will be in the legacy MVC proxied app. So what is the routing strategy configuration to use. With the default config from getting started I expected and routes that I have in my new .net6 MVC app to be used before being proxied, but that doesn't seem to be the case. Is there a way to specify which routes I do *NOT* want proxied? Or why doesn't it pick up the default route config
`{controller=Home}/{action=Index}/{id?}`

Secondly any thoughts or patterns regarding authentications and session info? I saw a video from School of Grok that does this creating API endpoints to serialize Principal and Claim info from proxy app to new app. Is this the recommended approach. Would be good if there was some examples showing some of these best practices for legacy app migration as I think more devs will be leverage YARP for that.

**Reply by Tratcher** [reply]
There's something about the default route `{controller=Home}/{action=Index}/{id?}` that seems to match at a lower priority than the proxy wildcard route `{**remainder}`. Still looking into that. In the meantime you can adjust the priority of the proxy route like this:
https://github.com/microsoft/reverse-proxy/discussions/1373#discussioncomment-1627540

Sharing auth and session between the apps are hard problems.
For auth it depends on the kind of auth. For session, it's almost impossible because the session designs are so different between the frameworks.

**Reply by pstatho** [reply]
Awesome! thanks for the quick reply. I confirm setting the Order to any value >0 in the default proxy route now displays new app routes and anything missing will get proxied. Intuitively the default MVC route `{controller=Home}/{action=Index}/{id?} ` is more specific the default **catch-all in the proxy config, but I'll leave that to you to investigate.

**Reply by Tratcher** [reply]
Ok, I got an explanation from a routing expert. It depends on who's creating the routes. MVC creates routes like `{controller=Home}/{action=Index}/{id?}`, starts at Order=1, and counts up in the order they're defined. Map* and YARP default to Order=0 and are higher priority than MVC routes.

**Reply by pstatho** [reply]
Ok so that explains why when I set the order of the YARP catch-all to 1 I start seeing the .net6 routing working. The order number is equally but controller/action is more specific and takes precedence. Thanks for the follow up
